base/slicearray.jl

```diff
@@ -0,0 +1,223 @@
+    _eachslice(A, dims, drop)
+end
+
+"""
+    eachrow(A::AbstractVecOrMat) <: AbstractVector
+
+Create a [`Rows`](@ref) object that is a vector of rows of matrix or vector `A`.
+Row slices are returned as `AbstractVector` views of `A`.
+
+See also [`eachcol`](@ref) and [`eachslice`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> S = eachrow(a)
+2-element Rows{Matrix{Int64}, Tuple{Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}}:
+ [1, 2]
+ [3, 4]
+
+julia> S[1]
+2-element view(::Matrix{Int64}, 1, :) with eltype Int64:
+ 1
+ 2
+```
+"""
+eachrow(A::AbstractMatrix) = _eachslice(A, (1,), true)
+eachrow(A::AbstractVector) = eachrow(reshape(A, size(A,1), 1))
+
+"""
+    eachcol(A::AbstractVecOrMat) <: AbstractVector
+
+Create a [`Columns`](@ref) object that is a vector of columns of matrix or vector `A`.
+Column slices are returned as `AbstractVector` views of `A`.
+
+See also [`eachrow`](@ref) and [`eachslice`](@ref).
+
+!!! compat "Julia 1.1"
+     This function requires at least Julia 1.1.
+
+!!! compat "Julia 1.7"
+     Prior to Julia 1.7, this returned an iterator.
+
+# Example
+
+```jldoctest
+julia> a = [1 2; 3 4]
+2×2 Matrix{Int64}:
+ 1  2
+ 3  4
+
+julia> S = eachcol(a)
+2-element Columns{Matrix{Int64}, Tuple{Base.OneTo{Int64}}, SubArray{Int64, 1, Matrix{Int64}, Tuple{Base.Slice{Base.OneTo{Int64}}, Int64}, true}}:
+ [1, 3]
+ [2, 4]
+
+julia> S[1]
+2-element view(::Matrix{Int64}, :, 1) with eltype Int64:
+ 1
+ 3
+```
+"""
+eachcol(A::AbstractMatrix) = _eachslice(A, (2,), true)
+eachcol(A::AbstractVector) = eachcol(reshape(A, size(A, 1), 1))
+
+"""
+    Rows{M,AX,S}
+
+A special case of [`Slices`](@ref) that is a vector of row slices of a matrix, as
+constructed by [`eachrow`](@ref).
+
+[`parent(S)`](@ref) can be used to get the underlying matrix.
+"""
+const Rows{P<:AbstractMatrix,AX,S<:AbstractVector} = Slices{P,Tuple{Int,Colon},AX,S,1}
+
+"""
+    Columns{M,AX,S}
+
+A special case of [`Slices`](@ref) that is a vector of column slices of a matrix, as
+constructed by [`eachcol`](@ref).
+
+[`parent(S)`](@ref) can be used to get the underlying matrix.
+"""
+const Columns{P<:AbstractMatrix,AX,S<:AbstractVector} = Slices{P,Tuple{Colon,Int},AX,S,1}
+
+
+IteratorSize(::Type{Slices{P,SM,AX,S,N}}) where {P,SM,AX,S,N} = HasShape{N}()
+axes(s::Slices) = s.axes
+size(s::Slices) = map(length, s.axes)
+
+@inline function _slice_index(s::Slices, c...)
+    return map(l -> l === (:) ? (:) : c[l], s.slicemap)
+end
+
+getindex(s::Slices{P,SM,AX,S,N}, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
+    view(s.parent, _slice_index(s, I...)...)
+setindex!(s::Slices{P,SM,AX,S,N}, val, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
+    s.parent[_slice_index(s, I...)...] = val
```

Comment:
I think it would be useful if the `@inbounds` context would be forwarded here:
```suggestion
Base.@propagate_inbounds getindex(s::Slices{P,SM,AX,S,N}, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
    view(s.parent, _slice_index(s, I...)...)
Base.@propagate_inbounds setindex!(s::Slices{P,SM,AX,S,N}, val, I::Vararg{Int,N}) where {P,SM,AX,S,N} =
    s.parent[_slice_index(s, I...)...] = val
```